Token.py

```diff
@@ -21,12 +22,160 @@ def __init__(
+            for w in kw:
+                for t in text_array:
+                    if re.search("^" + w, t) and len(text_array) < 10:
+                        return self.text
+
+    # returns the text if "total" or some variant is contained in text and group is fewer than 5 words
+    def get_total_label(self):
+        kw = ["total", "Total"]
+        if self.text:
+            for w in kw:
+                if w in self.text and len(self.text.split(" ")) < 5:
+                    return self.text
+
+    # returns string for description of number, eg. account number, invoice number
+    def get_num_label(self):
+        kw = ["no.", "no:", "number", "num", "No.", "No:"]
+        if self.text:
+            text_array = self.text.split(" ")
+            for w in kw:
+                if w in text_array and text_array.index(w) > 0:
+                    return text_array[text_array.index(w) - 1]
+
+    # returns a dictionary of {cur: <prefix> , value: <dollar amt> }
+    # eg. {cur: $ , value: 5.00 }
+    def get_currency(self):
+        currencies = ["$", "¥", "dollar", "SGD", "USD", "US$", "SG$", "$SG", "$US"]
+        out = {}
+        for cur in currencies:
+            if self.text and cur in self.text:
+                out["cur"] = cur
+                start = self.text.index(cur) + len(cur)
+                for i in range(start, len(self.text)):
+                    char = self.text[i]
+                    if not char.isdigit() and not (char in [".", ","]):
+                        break
+                try:
+                    out["val"] = float(self.text[start : i + 1])
+                    return out
+                except:
+                    return None
+
+    # checks if token is a date token
+    def get_dates(self):
+        # TODO parse range of dates and into date objects
+        dates = []
+        text = self.text
+        if type(text) is str:
+            month_names = set(
+                [
+                    "jan",
+                    "feb",
+                    "mar",
+                    "apr",
+                    "may",
+                    "jun",
+                    "jul",
+                    "aug",
+                    "sep",
+                    "oct",
+                    "nov",
+                    "dec",
+                ]
+            )
+
+            # checks for numerical months using regex
+            text_nospaces = text.replace(" ", "")
+            # matches d/dd or d-dd
+            re_date = re.search("\d[/|-]\d\d", text_nospaces)
+            if re_date:
+                dates.append(text_nospaces)
+
+            text_list = text.split(" ")
+
+            # if token has multiple words
+            if len(text_list) > 1:
+                # checks for named months
+                for index, word in enumerate(text_list):
+                    for month in month_names:
+                        if month in word.lower():
+                            date = []
```

Comment:
Wondering - is there a reason why we set date back to an empty list here?

Reply:
date and dates are different lists, we reset the date list every loop of month